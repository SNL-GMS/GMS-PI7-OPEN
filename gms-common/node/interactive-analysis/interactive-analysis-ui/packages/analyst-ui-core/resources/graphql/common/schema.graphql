# Common GraphQL schema definitions shared across the components of the
# API Gateway (e.g. waveform, workflow)

# A date/time Type
scalar Date

# Represents a time range
# type TimeRange {
#     startTime: Float!
#     endTime: Float!
# }

# Input verion of a time range
input TimeRange {
    startTime: Float!
    endTime: Float!
}

# Represents a location specified using latitude (degrees), longitude (degrees),
# and altitude (kilometers).
type Location {

    # Latitude in degrees
    latDegrees: Float!
    
    # Longitude in degrees
    lonDegrees: Float!

    # Elevation in kilometers
    elevationKm: Float!
}

input LocationInput {

    # Latitude in degrees
    latDegrees: Float!
    
    # Longitude in degrees
    lonDegrees: Float!

    # Elevation in kilometers
    elevationKm: Float!
}

# Enumerated list of source types used to compute distances to
enum DistanceSourceType {
    Event
    UserDefined
}

# Represents a distance measurement relative to a specified source location
type DistanceToSource {

    # The distance in kilometers to the source
    distanceKm: Float!

    # The source location
    sourceLocation: Location!

    # The type of the source the distance is measured to (e.g. and event)
    sourceType: DistanceSourceType!

    # Optional: the unique ID of the source object
    sourceId: String!
}

input DistanceToSourceInput {
    # The source location
    sourceLocation: LocationInput

    # The type of the source the distance is measured to (e.g. and event)
    sourceType: DistanceSourceType

    # Optional: the unique ID of the source object
    sourceId: String
}

# Represents a frequency range
type FrequencyBand {
    # Minimum frequency in Herz
    minFrequencyHz: Float!

    # Maximum frequency in Herz
    maxFrequencyHz: Float!
}

# Represents a frequency range input to a mutation
input FrequencyBandInput {
    # Minimum frequency in Herz
    minFrequencyHz: Float!

    # Maximum frequency in Herz
    maxFrequencyHz: Float!
}

# Enumeration representing the different types of stations in the monitoring network
enum StationType {

    Seismic3Component,
    Seismic1Component,
    SeismicArray,
    Hydroacoustic,
    HydroacousticArray,
    Infrasound,
    InfrasoundArray,
    Weather,
    UNKNOWN
}

# Provenance reference details for information added to the System
type InformationSource {

    # Unique ID for the source
    id: String!

    # Source organization
    organizationName: String!

    # The time associated with the information source
    referenceTime: Float!
}

# Enumeration of configuration status changes reflected in the configuration
enum ConfigurationInfoStatus {

    # Initial configuration
    Initial

    # The system's installed location for the item associated with the configuration was corrected
    CorrectedSystemLocation

    # The physical location of the ite associated with the configuration was changed
    MovedPhysicalLocation

    # The item associated with the configuration was decommissioned
    Decommissioned
}

# Provenance information associated with a configuration item in the system
type ConfigurationInfo {

    # The unique ID of the user who created the Configuration
    creatorId: String!

    # The time the configuration took effect in the real world (e.g. station came online)
    effectiveTime: Float!

    # The time the configuration was installed (and thus took effect) in the System
    installationTime: Float!

    # The status change reflected in the configuration values
    status: ConfigurationInfoStatus!

    # Comment associated with the configuration information
    comment: String
}

# Enumerated type of the actor (e.g. analyst, system) associated with a CreationInfo object
enum CreatorType {
    Analyst
    System
}

# Provenance information about the results of data processing completed by the System
# and actions completed by users of the system
type CreationInfo {
    
    # The unique ID of the creation info object
    id: String!

    # The time (seconds since Unix epoch) the associated processing result was created
    creationTime: Float!

    # The unique ID of the user who created the associated processing result
    creatorId: String!

    # The enumerated type of the creator (e.g. analyst, system)
    creatorType: CreatorType!

    # TODO
    # The processing step associated with the processing result
    # processingStep: ProcessingStep!

    # TODO
    # The software component associated with the processing result
    # softwareComponent: SoftwareComponent!
}

# Represents an alias for a station, channel, site, etc.
type Alias {

    # The alias name
    name: String!

    # Provenance information about the alias
    configurationInfo: ConfigurationInfo
}

# Represents calibration information associated with a waveform
type ProcessingCalibration {

    # Calibration factor
    factor: Float!

    # Calibration factor error
    factorError: Float!

    # Calibration period
    period: Float!

    # Calibration time shift
    timeShift: Float!
}

# Input version of calibration information associated with a waveform
input ProcessingCalibrationInput {

    # Calibration factor
    factor: Float!

    # Calibration factor error
    factorError: Float!

    # Calibration period
    period: Float!

    # Calibration time shift
    timeShift: Float!
}

# Represents a generic timeseries interface containing the fields common to the
# concrete timeseries data types defined in the system (e.g. Waveform, FkPowerSpectrum, DetectionFeatureMap)
interface Timeseries {
    # Unique identifier for the timeseries
    id: String!

    # The start date/time (seconds since epoch) of the timeseries
    startTime: Float!

    # The end date/time (seconds since epoch) of the timeseries
    endTime: Float!

    # The sample rate (Hz) of the samples in the timeseries
    sampleRate: Float!

    # The sample count in the timeseries
    sampleCount: Int!
}

# Represents a time-bounded segment of waveform samples output from a channel (derived or raw).
# Waveforms implement the Timeseries interface
type Waveform implements Timeseries {

    # Unique identifier for the waveform
    id: String!

    # The start date/time (seconds since epoch) of the waveform
    startTime: Float!

    # The end date/time (seconds since epoch) of the waveform
    endTime: Float!

    # The sample rate (Hz) of the samples in the waveform
    sampleRate: Float!

    # The sample count in the waveform
    sampleCount: Int!

    # The time-ordered collection of waveform sample values
    waveformSamples: [Float]

    # Calibration information for the waveform
    calibration: ProcessingCalibration
}

# Input version of a time-bounded segment of waveform samples output from a channel (derived or raw).
# Waveforms implement the Timeseries interface
input WaveformInput {

    # Unique identifier for the waveform
    id: String!

    # The start date/time (seconds since epoch) of the waveform
    startTime: Float!

    # The end date/time (seconds since epoch) of the waveform
    endTime: Float!

    # The sample rate (Hz) of the samples in the waveform
    sampleRate: Float!

    # The sample count in the waveform
    sampleCount: Int!

    # The time-ordered collection of waveform sample values
    waveformSamples: [Float]

    # Calibration information for the waveform
    calibration: ProcessingCalibrationInput
}

# The following queries are supported
type Query {
    # Dummy
    queryCount: Int
}

# The following mutations are supported
type Mutation {
    # Dummy
    mutationCount: Int
}

# The following subscriptions are supported
type Subscription {
    # Dummy
    subscriptionCount: Int
}