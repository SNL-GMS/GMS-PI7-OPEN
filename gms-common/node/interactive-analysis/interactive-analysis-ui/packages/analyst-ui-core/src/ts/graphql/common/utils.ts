/**
 * Helper function to remove the __typename from object.
 * Gateway GraphQL doesn't like it.
 * @param object Generic object to strip __typename out of
 * @returns new object copied from param with __typename stripped out of
 */
export function removeTypeName(object: any): any {
    const newObj = {};
    // tslint:disable-next-line:forin
    for (const key in object) {
        if (object[key] instanceof Object) {
            newObj[key] = removeTypeName(object[key]);
        } else if (key !== '__typename') {
            newObj[key] = object[key];
        }
    }
    return newObj;
}
