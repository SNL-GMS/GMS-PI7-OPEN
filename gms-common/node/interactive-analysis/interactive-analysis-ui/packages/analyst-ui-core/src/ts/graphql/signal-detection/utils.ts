import {
  AmplitudeMeasurementValue,
  FeatureMeasurement,
  FeatureMeasurementTypeName,
  InstantMeasurementValue,
  NumericMeasurementValue,
  PhaseTypeMeasurementValue,
  StringMeasurementValue
} from '~graphql/signal-detection/types';

/**
 * Searches Feature Measurements for the desired Feature Measurement
 * @param featureMeasurements List of feature measurements
 * @param featureMeasurementType Enum of desired Feature Measurement desired
 * 
 * @returns FeatureMeasurement or undefined if not found
 */
export function findFeatureMeasurement(
  featureMeasurements: FeatureMeasurement[], featureMeasurementType: FeatureMeasurementTypeName):
  FeatureMeasurement | undefined {
  if (featureMeasurements && featureMeasurementType) {
    return featureMeasurements.find(fm => fm.featureMeasurementType === featureMeasurementType);
  }
  return undefined;
}

/**
 * Searches Feature Measurements for the ArrivalTime Feature Measurement
 * @param featureMeasurements List of feature measurements
 * 
 * @returns ArrivalTime FeatureMeasurement or undefined if not found
 */
export function findArrivalTimeFeatureMeasurement(
  featureMeasurements: FeatureMeasurement[]):
  FeatureMeasurement | undefined {
  return findFeatureMeasurement(featureMeasurements, FeatureMeasurementTypeName.ARRIVAL_TIME);
}

/**
 * Searches Feature Measurements for the ArrivalTime Feature Measurement Value
 * @param featureMeasurements List of feature measurements
 * 
 * @returns ArrivalTime FeatureMeasurementValue or undefined if not found
 */
export function findArrivalTimeFeatureMeasurementValue(
  featureMeasurements: FeatureMeasurement[]): InstantMeasurementValue | undefined {
  const fm = findArrivalTimeFeatureMeasurement(featureMeasurements);
  return (fm) ? fm.measurementValue as InstantMeasurementValue : undefined;
}

/**
 * Searches Feature Measurements for the Azimuth Feature Measurement
 * @param featureMeasurements List of feature measurements
 * 
 * @returns Azimuth FeatureMeasurement or undefined if not found
 */
export function findAzimthFeatureMeasurement(
  featureMeasurements: FeatureMeasurement[]):
  FeatureMeasurement | undefined {
  const azimuthList: FeatureMeasurementTypeName[] = [FeatureMeasurementTypeName.RECEIVER_TO_SOURCE_AZIMUTH,
      FeatureMeasurementTypeName.SOURCE_TO_RECEIVER_AZIMUTH];
  // Search FeatureMeasurements to find which type of Azimuth was supplied
  return featureMeasurements.find(fm => azimuthList.find(
      azTypeName => azTypeName === fm.featureMeasurementType) !== undefined);
}

/**
 * Searches Feature Measurements for the Azimuth Feature Measurement Value
 * @param featureMeasurements List of feature measurements
 * 
 * @returns Azimuth FeatureMeasurementValue or undefined if not found
 */
export function findAzimthFeatureMeasurementValue(
  featureMeasurements: FeatureMeasurement[]): NumericMeasurementValue | undefined {
  const fm = findAzimthFeatureMeasurement(featureMeasurements);
  return (fm) ? fm.measurementValue as NumericMeasurementValue : undefined;
}

/**
 * Searches Feature Measurements for the Slowness Feature Measurement
 * @param featureMeasurements List of feature measurements
 * 
 * @returns Slowness FeatureMeasurement or undefined if not found
 */
export function findSlownessFeatureMeasurement(
  featureMeasurements: FeatureMeasurement[]): FeatureMeasurement | undefined {
  return findFeatureMeasurement(featureMeasurements, FeatureMeasurementTypeName.SLOWNESS);
}

/**
 * Searches Feature Measurements for the Slowness Feature Measurement Value
 * @param featureMeasurements List of feature measurements
 * 
 * @returns Slowness FeatureMeasurementValue or undefined if not found
 */
export function findSlownessFeatureMeasurementValue(
  featureMeasurements: FeatureMeasurement[]): NumericMeasurementValue | undefined {
  const fm = findSlownessFeatureMeasurement(featureMeasurements);
  return (fm) ? fm.measurementValue as NumericMeasurementValue : undefined;
}

/**
 * Searches Feature Measurements for the Amplitude Feature Measurement
 * @param featureMeasurements List of feature measurements
 * 
 * @returns Phase FeatureMeasurement or undefined if not found
 */
export function findAmplitudeFeatureMeasurement(
  featureMeasurements: FeatureMeasurement[]): FeatureMeasurement | undefined {
  const amplitudeList: FeatureMeasurementTypeName[] = [FeatureMeasurementTypeName.AMPLITUDE,
  FeatureMeasurementTypeName.AMPLITUDE_A5_OVER_2, FeatureMeasurementTypeName.AMPLITUDE_ALR_OVER_2,
  FeatureMeasurementTypeName.AMPLITUDE_ANL_OVER_2, FeatureMeasurementTypeName.AMPLITUDEh_ALR_OVER_2];

  // Search FeatureMeasurements to find which type of Amplitude was supplied
  return featureMeasurements.find(fm => amplitudeList.find(
    ampTypeName => ampTypeName === fm.featureMeasurementType) !== undefined);
}

/**
 * Searches Feature Measurements for the Amplitude Feature Measurement Value
 * @param featureMeasurements List of feature measurements
 * 
 * @returns Phase FeatureMeasurementValue or undefined if not found
 */
export function findAmplitudeFeatureMeasurementValue(
  featureMeasurements: FeatureMeasurement[]): AmplitudeMeasurementValue | undefined {
  const fm = findAmplitudeFeatureMeasurement(featureMeasurements);
  return (fm) ? fm.measurementValue as AmplitudeMeasurementValue : undefined;
}

/**
 * Searches Feature Measurements for the Phase Feature Measurement
 * @param featureMeasurements List of feature measurements
 * 
 * @returns Phase FeatureMeasurement or undefined if not found
 */
export function findPhaseFeatureMeasurement(
  featureMeasurements: FeatureMeasurement[]): FeatureMeasurement | undefined {
  return findFeatureMeasurement(featureMeasurements, FeatureMeasurementTypeName.PHASE);
}

/**
 * Searches Feature Measurements for the Phase Feature Measurement Value
 * @param featureMeasurements List of feature measurements
 * 
 * 
 * @returns Phase FeatureMeasurementValue or undefined if not found
 */
export function findPhaseFeatureMeasurementValue(
  featureMeasurements: FeatureMeasurement[]): PhaseTypeMeasurementValue | undefined {
  const fm = findPhaseFeatureMeasurement(featureMeasurements);
  return (fm) ? fm.measurementValue as PhaseTypeMeasurementValue : undefined;
}

/**
 * Searches Feature Measurements for the Phase Feature Measurement Value
 * @param featureMeasurements List of feature measurements
 * 
 * @returns Phase FeatureMeasurementValue or undefined if not found
 */
export function findFilteredBeamFeatureMeasurement(
  featureMeasurements: FeatureMeasurement[], filterDefinitionId: string): FeatureMeasurement| undefined {
    if (featureMeasurements) {
      const filteredBeamFM = featureMeasurements.find(fm =>
        fm.featureMeasurementType === FeatureMeasurementTypeName.FILTERED_BEAM &&
        (fm.measurementValue as StringMeasurementValue).strValue === filterDefinitionId);
      return filteredBeamFM;
    }
    return undefined;
}
