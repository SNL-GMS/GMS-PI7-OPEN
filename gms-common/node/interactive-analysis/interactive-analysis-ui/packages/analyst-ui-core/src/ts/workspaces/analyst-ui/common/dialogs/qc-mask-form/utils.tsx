/**
 * A helper utility that builds buttons for the Table view in the qcMaskDialogBox
 */
import { Button, Icon, Intent } from '@blueprintjs/core';
import { Column, Table, TimeUtil } from '@gms/ui-core-components';
import * as classNames from 'classnames';
import * as React from 'react';

/**
 * Builds a modify or reject button
 * 
 * @param params Table (ag-grid) parameters
 * 
 * @returns a JSX.Element or null
 */
export function modifyButton(params): JSX.Element | null {

  const modifyIcon = (<Icon icon="edit" title={false} />);
  const viewIcon = (<Icon icon="eye-open" title={false} />);

  return (
    !params.value.disabled ?
      (
        <Button
          onClick={e => {
            e.stopPropagation();
            params.value.onClick(e.clientX, e.clientY, params);
          }}
          className="qc-mask-history-table__button"
          icon={modifyIcon}
          small={true}
          minimal={true}
          title={'modify'}

        />
      )
      : (
        <Button
          onClick={e => {
            e.stopPropagation();
            params.value.onClick(e.clientX, e.clientY, params);
          }}
          className="qc-mask-history-table__button"
          icon={viewIcon}
          title={'view'}
          small={true}
          minimal={true}

        />)
  );
}
/**
 * Builds a button for selecting qc masks
 * 
 * @param Table (ag-grid) parameters
 * 
 * @returns a JSX.Element or null
 */
export function selectButton(params) {
  const selectIcon = (<Icon icon="select" title={false} />);

  return (
    !params.value.disabled ?
      (
        <Button
          onClick={e => {
            e.stopPropagation();
            params.value.onClick(e.clientX, e.clientY, params);
          }}
          className="qc-mask-history-table__button"
          icon={selectIcon}
          small={true}
          minimal={true}
          title={'select'}
        />)
      : null
  );
}
/**
 * Builds a button for opening the qc mask reject dialog
 * 
 * @param Table (ag-grid) parameters
 * 
 * @returns a JSX.Element or null
 */
export function rejectButton(params) {
  const rejectIcon = (<Icon icon="cross" title={false} />);

  return (
    !params.value.disabled ?
      (
        <Button
          onClick={e => {
            e.stopPropagation();
            params.value.onClick(e.clientX, e.clientY, params);
          }}
          className="qc-mask-history-table__button qc-mask-history-table__button--reject"
          icon={rejectIcon}
          small={true}
          minimal={true}
          title={'reject'}
          intent={Intent.DANGER}
        />)
      : null
  );
}

/**
 * Column definitions for the overlapping mask table.
 */
const OVERLAPPING_MASKS_COLUMN_DEFINITIONS: Column[] = [
  {
    headerName: '',
    field: 'select',
    cellStyle: { 'text-align': 'left', 'vertical-align': 'middle' },
    width: 25,
    cellRendererFramework: selectButton,
    enableCellChangeFlash: true

  },
  {
    headerName: '',
    field: 'modify',
    cellStyle: { 'text-align': 'left', 'vertical-align': 'middle' },
    width: 25,
    cellRendererFramework: modifyButton,
    enableCellChangeFlash: true
  },
  {
    headerName: '',
    field: 'reject',
    cellStyle: { 'text-align': 'left', 'vertical-align': 'middle' },
    width: 25,
    cellRendererFramework: rejectButton,
    enableCellChangeFlash: true

  },
  {
    headerName: '',
    field: 'color',
    cellStyle: { 'text-align': 'left', 'vertical-align': 'middle' },
    width: 30,
    cellRendererFramework: e => (
      <div
        style={{
          height: '10px',
          width: '20px',
          backgroundColor: e.data.color,
          marginTop: '4px'
        }}
      />
    ),
    enableCellChangeFlash: true
  },
  {
    headerName: 'Creation time',
    field: 'timestamp',
    cellStyle: { 'text-align': 'left' },
    width: 170,
    enableCellChangeFlash: true,
    valueFormatter: e => TimeUtil.toString(e.data.creationTime)
  },
  {
    headerName: 'Category',
    field: 'category',
    cellStyle: { 'text-align': 'left' },
    width: 130,
    enableCellChangeFlash: true
  },
  {
    headerName: 'Type',
    field: 'type',
    cellStyle: { 'text-align': 'left' },
    width: 130,
    enableCellChangeFlash: true
  },
  {
    headerName: 'Start time',
    field: 'startTime',
    cellStyle: { 'text-align': 'left' },
    width: 170,
    enableCellChangeFlash: true,
    valueFormatter: e => TimeUtil.toString(e.data.startTime)
  },
  {
    headerName: 'End time',
    field: 'endTime',
    cellStyle: { 'text-align': 'left' },
    width: 170,
    enableCellChangeFlash: true,
    valueFormatter: e => TimeUtil.toString(e.data.endTime)
  },
  {
    headerName: 'Author',
    field: 'author',
    cellStyle: { 'text-align': 'left' },
    width: 75,
    enableCellChangeFlash: true
  },
  {
    headerName: 'Rationale',
    field: 'rationale',
    cellStyle: { 'text-align': 'left' },
    width: 300,
    enableCellChangeFlash: true
  }
];

/**
 * Render the Mask table (used for history and overlapping masks).
 * 
 * @param tableProps a set of props used by table
 * 
 * @returns a JSX.Element with a rendered table
 */
export const renderOverlappingMaskTable = (tableProps: {}) => (
  <div
    className={classNames('ag-dark', 'qc-mask-overlapping-table')}
  >
    <div style={{ flex: '1 1 auto', position: 'relative', minHeight: '150px' }}>
      <div className="max">
        <Table
          columnDefs={OVERLAPPING_MASKS_COLUMN_DEFINITIONS}
          getRowNodeId={node => node.id}
          rowSelection="single"
          {...tableProps}
        />
      </div>
    </div>
  </div>
);
