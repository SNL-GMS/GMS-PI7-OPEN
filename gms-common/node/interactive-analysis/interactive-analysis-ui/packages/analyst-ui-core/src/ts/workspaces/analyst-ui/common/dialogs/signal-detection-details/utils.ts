/**
 * Reusable formatter for time uncertainty 
 * 
 * @param unc the uncertainty value in seconds
 * 
 * @returns a string with two trailing digits
 */
export function formatUncertainty(unc: number): string {
  if (!unc) {
    return '';
  }
  return `${unc.toFixed(2)} s`;
}
