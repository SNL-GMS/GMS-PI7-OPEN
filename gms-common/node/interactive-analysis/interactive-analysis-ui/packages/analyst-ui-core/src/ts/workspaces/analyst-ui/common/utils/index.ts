import * as EventUtils from './event-util';
import * as SignalDetectionUtils from './signal-detection-util';
export { EventUtils, SignalDetectionUtils};
