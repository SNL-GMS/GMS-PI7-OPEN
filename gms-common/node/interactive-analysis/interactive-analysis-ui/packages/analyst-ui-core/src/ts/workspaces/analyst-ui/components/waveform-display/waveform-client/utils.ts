import { ChannelSegmentTypes, WaveformTypes } from '~graphql/';

/**
 * Creates a zeroe'd out array of waveform data with given time interva
 * 
 * @param startTimeSecs start of the array
 * @param endTimeSecs end of the array
 * 
 * @returns float32 array with values of all zero
 */
export function createZeroArray(startTimeSecs: number, endTimeSecs: number): number[] {
    return Array.from(
        new Float32Array(
          WaveformTypes.DEFAULT_SAMPLE_RATE * (endTimeSecs - startTimeSecs) + 1
        ).fill(0)
    );
}
/**
 * Creates a channel segment object with all zero values for waveform data
 * 
 * @param cId channelId to create mock segment for
 * @param fId filterId to create mock segment for
 * @param startTimeSecs start time for mock segment time interval
 * @param endTimeSecs end time for mock segment time interval
 * @param data data to be used as wf data
 */
export const createZeroDataChannelSegment = (
    cId: string,
    fId: string,
    startTimeSecs: number,
    endTimeSecs: number,
    data: number[]
  ): ChannelSegmentTypes.ChannelSegment<WaveformTypes.Waveform> => {
    const ts: WaveformTypes.Waveform = {
      startTime: startTimeSecs,
      sampleRate: WaveformTypes.DEFAULT_SAMPLE_RATE,
      sampleCount: 1,
      values: data
    };
    return ({
      id: `${cId}${fId}`,
      type: ChannelSegmentTypes.ChannelSegmentType.RAW,
      channelId: cId,
      startTime: startTimeSecs,
      endTime: endTimeSecs,
      timeseriesType: ChannelSegmentTypes.TimeSeriesType.WAVEFORM,
      timeseries: [ts]
  });
};

/**
 * Creates a channel segment object with all zero values for waveform data
 * 
 * @param cId channelId to create mock segment for
 * @param fId filterId to create mock segment for
 * @param startTimeSecs start time for mock segment time interval
 * @param endTimeSecs end time for mock segment time interval
 * @param data data to be used as wf data
 */
export const createZeroDataFilteredChannelSegment = (
    cId: string,
    fId: string,
    startTimeSecs: number,
    endTimeSecs: number,
    data: number[]
  ): WaveformTypes.FilteredChannelSegment => {
    const ts: WaveformTypes.Waveform = {
      startTime: startTimeSecs,
      sampleRate: WaveformTypes.DEFAULT_SAMPLE_RATE,
      sampleCount: 1,
      values: data
    };
    return ({
      id: `${cId}${fId}`,
      type: ChannelSegmentTypes.ChannelSegmentType.FILTER,
      wfFilterId: fId,
      channelId: cId,
      sourceChannelId: cId,
      startTime: startTimeSecs,
      endTime: endTimeSecs,
      timeseriesType: ChannelSegmentTypes.TimeSeriesType.WAVEFORM,
      timeseries: [ts]
    });
  };
