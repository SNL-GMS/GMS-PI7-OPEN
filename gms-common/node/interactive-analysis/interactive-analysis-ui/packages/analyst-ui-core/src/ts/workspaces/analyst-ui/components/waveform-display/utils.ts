import { FeaturePrediction } from '~graphql/event/types';
import { InstantMeasurementValue } from '~graphql/signal-detection/types';

export interface Offset {
  channelId: string;
  offset: number;
}

/**
 * Calculate offsets based on station with earliest arrival.
 */
export const calculateOffsets = (
  featurePredictions: FeaturePrediction[], phaseToOffset: string): Offset[] => {

  if (!featurePredictions || featurePredictions.length <= 0) {
    return [];
  }
  const filteredPredictions: FeaturePrediction[] =
    featurePredictions.filter(fp => fp.phase === phaseToOffset);

  // If no FPs were found for the phase selected, return nothing
  if (filteredPredictions.length <= 0) {
    return [];
  }

  filteredPredictions.sort((a, b) => {
    const aValue = (a.predictedValue as InstantMeasurementValue).value;
    const bValue = (b.predictedValue as InstantMeasurementValue).value;
    return aValue - bValue;
  });
  const baseTime: number = (filteredPredictions[0].predictedValue as InstantMeasurementValue).value;

  const offsets: Offset[] = [];

  filteredPredictions.forEach(sortedFeaturePrediction => {
    const offset: Offset = {
      channelId: sortedFeaturePrediction.channelId,
      offset: baseTime - (sortedFeaturePrediction.predictedValue as InstantMeasurementValue).value
    };
    offsets.push(offset);
  });

  return offsets;
  };
