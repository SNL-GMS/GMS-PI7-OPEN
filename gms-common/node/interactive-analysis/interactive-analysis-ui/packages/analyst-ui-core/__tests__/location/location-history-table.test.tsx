import * as React from 'react';
import { EventUtils } from '~analyst-ui/common/utils';
import {
  LocationHistory,
} from '~analyst-ui/components/location/components/location-history';
// tslint:disable-next-line:max-line-length
import * as LocationData from '../__data__/location-data';

// tslint:disable-next-line:no-implicit-dependencies no-var-requires no-require-imports
const Enzyme = require('enzyme');

it('Location History Table Renders a single location entry', () => {
  const wrapper = Enzyme.shallow(
    <LocationHistory
      event={LocationData.event}
      selectedLocationSolutionSetId={''}
      selectedLocationSolutionId={''}
      setSelectedLSSAndLS={undefined}
    />
  );
  expect(wrapper)
    .toMatchSnapshot();
});

// Unit test for checking that the correct, configured preferred restraints is picked
describe('When location is set to save, correct preferred location is set', async () => {
  it('The osd event hypothesis should match expected result', async () => {
    const preferredId =
      EventUtils.getPreferredLocationIdFromEventHyp(LocationData.eventHypothesisWithLocationSets);
    const hardCodedCorrectId = '3a06fac7-46ad-337e-a8da-090a1cc801a1';
    expect(preferredId)
    .toEqual(hardCodedCorrectId);
  });

});
