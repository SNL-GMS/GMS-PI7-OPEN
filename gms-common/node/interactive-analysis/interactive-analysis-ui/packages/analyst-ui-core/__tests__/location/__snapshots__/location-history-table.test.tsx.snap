// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Location History Table Renders a single location entry 1`] = `
<div
  className="ag-theme-dark table-container"
>
  <div
    className="list-wrapper"
  >
    <div
      className="max"
    >
      <Table
        autoGroupColumnDef={
          Object {
            "cellRendererFramework": [Function],
            "comparator": [Function],
            "field": "save",
            "headerName": "Save",
            "headerTooltip": "Save this location set when the event is saved",
            "width": 65,
          }
        }
        columnDefs={
          Array [
            Object {
              "cellRendererFramework": [Function],
              "field": "locationSetId",
              "headerName": "ASBDUSA",
              "hide": true,
              "rowGroup": true,
              "sortable": true,
            },
            Object {
              "cellStyle": Object {
                "text-align": "left",
              },
              "enableCellChangeFlash": true,
              "field": "locType",
              "headerName": "Type",
              "headerTooltip": "Event Location Type (Standard or Master)",
              "resizable": true,
              "width": 90,
            },
            Object {
              "cellStyle": Object {
                "text-align": "right",
              },
              "enableCellChangeFlash": true,
              "field": "lat",
              "headerName": "Lat (°)",
              "headerTooltip": "Latitude of event location",
              "resizable": true,
              "width": 70,
            },
            Object {
              "cellStyle": Object {
                "text-align": "right",
              },
              "enableCellChangeFlash": true,
              "field": "lon",
              "headerName": "Lon (°)",
              "headerTooltip": "Longitude of event location",
              "resizable": true,
              "width": 70,
            },
            Object {
              "cellStyle": Object {
                "text-align": "right",
              },
              "enableCellChangeFlash": true,
              "field": "depth",
              "headerName": "Depth (km)",
              "headerTooltip": "Depth of event location",
              "resizable": true,
              "width": 90,
            },
            Object {
              "cellStyle": Object {
                "text-align": "right",
              },
              "enableCellChangeFlash": true,
              "field": "time",
              "headerName": "Time",
              "headerTooltip": "Time of event",
              "resizable": true,
              "width": 210,
            },
            Object {
              "cellStyle": Object {
                "text-align": "left",
              },
              "field": "restraint",
              "headerName": "Restraint",
              "headerTooltip": "Depth restraint of location calculation",
              "resizable": true,
              "width": 140,
            },
            Object {
              "cellStyle": Object {
                "text-align": "right",
              },
              "enableCellChangeFlash": true,
              "field": "smajax",
              "headerName": "Semi Major (km)",
              "headerTooltip": "Semi-major axis length of error ellipse",
              "resizable": true,
              "width": 122,
            },
            Object {
              "cellStyle": Object {
                "text-align": "right",
              },
              "enableCellChangeFlash": true,
              "field": "sminax",
              "headerName": "Semi Minor (km)",
              "headerTooltip": "Semi-minor axis length of error ellipse",
              "resizable": true,
              "width": 122,
            },
            Object {
              "cellStyle": Object {
                "text-align": "right",
              },
              "enableCellChangeFlash": true,
              "field": "strike",
              "headerName": "Strike (°)",
              "headerTooltip": "Strike of major axis of error ellipse",
              "resizable": true,
              "width": 70,
            },
            Object {
              "cellStyle": Object {
                "text-align": "right",
              },
              "enableCellChangeFlash": true,
              "field": "stdev",
              "headerName": "Std Dev",
              "headerTooltip": "Standard deviation of the observed",
              "resizable": true,
              "width": 70,
            },
            Object {
              "cellStyle": Object {
                "text-align": "right",
              },
              "enableCellChangeFlash": true,
              "field": "count",
              "headerName": "Count",
              "hide": true,
              "resizable": true,
              "sortable": true,
              "width": 95,
            },
            Object {
              "cellRendererFramework": [Function],
              "cellStyle": Object {
                "display": "flex",
                "justify-content": "center",
              },
              "field": "preferred",
              "headerName": "Preferred",
              "headerTooltip": "Mark an event location as preferred",
              "width": 80,
            },
          ]
        }
        context={Object {}}
        getNodeChildDetails={[Function]}
        getRowNodeId={[Function]}
        getRowStyle={[Function]}
        groupMultiAutoColumn={true}
        onCellClicked={[Function]}
        onGridReady={[Function]}
        onRowClicked={[Function]}
        rowData={
          Array [
            Object {
              "count": 0,
              "depth": "0.000",
              "depthRestraintType": "UNRESTRAINED",
              "id": "1",
              "isFirstInLSSet": true,
              "isLastInLSSet": true,
              "isLocationSolutionSetPreferred": true,
              "lat": "44.949",
              "locType": "Standard",
              "locationSetId": "5fa862f1-2e2d-4d87-bc66-00177de18c5e",
              "locationSolutionId": "3a06fac7-46ad-337e-a8da-090a1cc801a1",
              "lon": "-106.384",
              "preferred": true,
              "restraint": "Unrestrained",
              "selectedLocationSolutionSetId": "",
              "setPreferred": [Function],
              "setToSave": [Function],
              "smajax": "59.714",
              "sminax": "32.373",
              "stdev": "0.000",
              "strike": "137.570",
              "time": "2010/05/20T22:01:30.238000",
            },
          ]
        }
        rowSelection="none"
        suppressContextMenu={true}
      />
    </div>
  </div>
</div>
`;
